### PR TITLE
feat: add model implementation for UTF8 enc/dec

### PR DESCRIPTION
**Pull request by digama0**
- [x] Depends on: #3958 
- [x] Depends on: #3960

This makes the UTF-8 encode and decode functions have lean definitions, so that we can prove properties about them downstream.

### CODE REVIEW

**Comment by semorrison** [comment]
Could you add a basic test file using `test_extern`, e.g. starting with

```
import Lean.Util.TestExtern

instance : BEq ByteArray where
  beq x y := x.data == y.data

test_extern String.toUTF8 "$£€𐍈"
```
and then similarly for the other functions?

**Review by semorrison — APPROVED** [review]

**Comment by semorrison** [comment]
LGTM, but I'm not the one to merge changes to UTF8 handling!

File: tests/lean/run/utf8英語.lean [hunk 1]
```diff
@@ -1,3 +1,47 @@
...
+instance : BEq ByteArray where
+  beq x y := x.data == y.data
+
+test_extern String.toUTF8 ""
+test_extern String.toUTF8 "\x00"
+test_extern String.toUTF8 "$£€𐍈"
+
+macro "test_extern'" t:term " => " v:term : command =>
+  `(test_extern $t
+    #guard $t == $v)
```
Comment (nomeata) [inline tests/lean/run/utf8英語.lean:12]: Can we have this next to `test_extern` with a more helpful name? Maybe
`guard_extern`? Or simply `test_extern` and using `=>` to enable the additional guard?

File: src/Init/Data/String/Extra.lean [hunk 2]
```diff
@@ -17,13 +17,69 @@ def toNat! (s : String) : Nat :=
...
+  let c ← a[i]?
+  if c &&& 0x80 == 0 then
+    some ⟨c.toUInt32, .inl (Nat.lt_trans c.1.2 (by decide))⟩
+  else if c &&& 0xe0 == 0xc0 then
+    let c1 ← a[i+1]?
+    guard (c1 &&& 0xc0 == 0x80)
+    let r := ((c &&& 0x1f).toUInt32 <<< 6) ||| (c1 &&& 0x3f).toUInt32
+    guard (0x80 ≤ r)
+    -- FIXME: this check is not needed
+    if h : r < 0xd800 then some ⟨r, .inl h⟩ else none
```
Comment (nomeata) [inline src/Init/Data/String/Extra.lean:30]: Why is it not? Why is it still there then?

File: tests/lean/run/utf8英語.lean (hunk 1 above)
Comment (nomeata) [inline tests/lean/run/utf8英語.lean:12]: Should we have this next to `test_extern` with a more helpful name? Maybe
`guard_extern`? Or simply use `test_extern` and look for the `=>` syntax to enable the additional guard?

**Review by nomeata — COMMENTED** [review]
Good stuff, minor comments only.

File: src/Init/Data/String/Extra.lean (hunk 2 above)
Comment (digama0) [reply src/Init/Data/String/Extra.lean:30]: it's possible to prove from the previous line that `r` is at most 2^11, which is less than 0xd800. But proving those kind of facts about bitwise ops is not easy, especially in this very early file.

File: tests/lean/run/utf8英語.lean (hunk 1 above)
Comment (digama0) [reply tests/lean/run/utf8英語.lean:12]: I tried to sneak this feature request in https://github.com/leanprover/lean4/pull/3075#issuecomment-2068637377 . It's a separate PR though in any case.

File: src/Init/Data/String/Extra.lean (hunk 2 above)
Comment (nomeata) [reply src/Init/Data/String/Extra.lean:30]: Ah, now I get it. How about

```suggestion
    -- We could prove h from the definition of r, but it is too tedious,
    -- so for now we just check it again.
    if h : r < 0xd800 then some ⟨r, .inl h⟩ else none
```

Comment (digama0) [reply src/Init/Data/String/Extra.lean:30]:
```suggestion
    -- TODO: Prove h from the definition of r once we have the necessary lemmas
    if h : r < 0xd800 then some ⟨r, .inl h⟩ else none
```

File: src/Init/Data/String/Extra.lean [hunk 3]
```diff
@@ -17,13 +17,69 @@ def toNat! (s : String) : Nat :=
...
+  else if c &&& 0xf0 == 0xe0 then
+    let c1 ← a[i+1]?
+    let c2 ← a[i+2]?
+    guard (c1 &&& 0xc0 == 0x80 && c2 &&& 0xc0 == 0x80)
+    let r :=
+      ((c &&& 0x0f).toUInt32 <<< 12) |||
+      ((c1 &&& 0x3f).toUInt32 <<< 6) |||
+      (c2 &&& 0x3f).toUInt32
+    guard (0x800 ≤ r)
+    -- FIXME: the `r < 0x110000` check is not needed
```
Comment (nomeata) [inline src/Init/Data/String/Extra.lean:40]:
```suggestion
    -- TODO: Prove h from the definition of r once we have the necessary lemmas
```

**Review by nomeata — APPROVED** [review]
Looks good. `test_extern` improvement can be generalized in a separate PR. Happy to merge if you are happy with this.

File: src/Init/Data/String/Extra.lean (hunk 3 above)
Comment (digama0) [reply src/Init/Data/String/Extra.lean:40]:
```suggestion
    -- TODO: Prove `r < 0x110000` from the definition of r once we have the necessary lemmas
```

**Pull request merged**